File: python/src/exercises/task09/number_formatting_exercises.py
"""
Task 9: Number Formatting Exercises

This module contains exercises for learning number formatting operations in Python.
Students will implement methods that demonstrate:
- Decimal formatting with specific precision
- Currency formatting for different locales
- Percentage formatting and calculations
- Scientific notation formatting
- Number parsing and validation
- Custom number format patterns

Learning Objectives:
- Master Python's number formatting methods
- Understand locale-specific number formatting
- Learn to format numbers for different contexts
- Practice number parsing and validation
- Handle different number formats and patterns
- Work with currency and percentage representations

Author: ZipCode Cohort
"""

from typing import Optional, Union
import locale
import math
import re


class NumberFormattingExercises:
    """
    Number formatting exercises for learning Python number operations.
    """

    def format_decimal(self, number: float, decimal_places: int) -> str:
        """
        Format a decimal number with a specific number of decimal places.
        Use string formatting to control precision.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_decimal(123.456789, 2)
        '123.46'
        >>> exercises.format_decimal(7.1, 3)
        '7.100'
        >>> exercises.format_decimal(42, 0)
        '42'
        >>> exercises.format_decimal(3.999, 1)
        '4.0'
        
        Args:
            number: Number to format
            decimal_places: Number of decimal places to show
            
        Returns:
            Formatted decimal string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_with_thousands_separator(self, number: float) -> str:
        """
        Format a number with thousands separators (commas).
        Use appropriate formatting to add commas for readability.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_with_thousands_separator(1234567.89)
        '1,234,567.89'
        >>> exercises.format_with_thousands_separator(1000)
        '1,000'
        >>> exercises.format_with_thousands_separator(42.5)
        '42.5'
        >>> exercises.format_with_thousands_separator(999)
        '999'
        
        Args:
            number: Number to format
            
        Returns:
            Number string with thousands separators
        """
        raise NotImplementedError("Method not implemented yet")

    def format_as_currency(self, amount: float) -> str:
        """
        Format a number as currency using default formatting.
        Use appropriate currency formatting with dollar sign and 2 decimal places.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_as_currency(123.45)
        '$123.45'
        >>> exercises.format_as_currency(0.99)
        '$0.99'
        >>> exercises.format_as_currency(1000)
        '$1,000.00'
        >>> exercises.format_as_currency(0)
        '$0.00'
        
        Args:
            amount: Amount to format as currency
            
        Returns:
            Formatted currency string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_as_currency_with_locale(self, amount: float, locale_name: str) -> str:
        """
        Format a number as currency using a specific locale.
        Use locale-specific currency formatting.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_as_currency_with_locale(123.45, 'en_US')
        '$123.45'
        >>> exercises.format_as_currency_with_locale(123.45, 'en_GB')
        '£123.45'
        >>> exercises.format_as_currency_with_locale(123.45, 'de_DE')
        '123,45 €'
        
        Args:
            amount: Amount to format as currency
            locale_name: Locale name to use for formatting
            
        Returns:
            Formatted currency string in the specified locale
        """
        raise NotImplementedError("Method not implemented yet")

    def format_as_percentage(self, value: float, decimal_places: int) -> str:
        """
        Format a decimal value as a percentage.
        Convert decimal to percentage (multiply by 100) and add % symbol.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_as_percentage(0.1234, 2)
        '12.34%'
        >>> exercises.format_as_percentage(0.5, 1)
        '50.0%'
        >>> exercises.format_as_percentage(1.0, 0)
        '100%'
        >>> exercises.format_as_percentage(0.001, 3)
        '0.100%'
        
        Args:
            value: Decimal value (0.5 = 50%)
            decimal_places: Number of decimal places to show
            
        Returns:
            Formatted percentage string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_in_scientific_notation(self, number: float, decimal_places: int) -> str:
        """
        Format a number in scientific notation.
        Use exponential notation with specified decimal places.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_in_scientific_notation(1234567, 2)
        '1.23e+06'
        >>> exercises.format_in_scientific_notation(0.000123, 3)
        '1.230e-04'
        >>> exercises.format_in_scientific_notation(42, 1)
        '4.2e+01'
        >>> exercises.format_in_scientific_notation(1, 0)
        '1e+00'
        
        Args:
            number: Number to format
            decimal_places: Number of decimal places in mantissa
            
        Returns:
            Scientific notation string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_with_leading_zeros(self, number: int, total_width: int) -> str:
        """
        Format a number with leading zeros to a specific total width.
        Use string formatting to add leading zeros.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_with_leading_zeros(42, 5)
        '00042'
        >>> exercises.format_with_leading_zeros(123, 6)
        '000123'
        >>> exercises.format_with_leading_zeros(7, 3)
        '007'
        >>> exercises.format_with_leading_zeros(1000, 3)
        '1000'
        
        Args:
            number: Integer to format
            total_width: Desired total width
            
        Returns:
            Number string with leading zeros
        """
        raise NotImplementedError("Method not implemented yet")

    def parse_float(self, number_string: str) -> float:
        """
        Parse a string to extract a float value.
        Handle various number formats and return the parsed value.
        Return 0.0 if parsing fails.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.parse_float('123.45')
        123.45
        >>> exercises.parse_float('1,234.56')
        1234.56
        >>> exercises.parse_float('$123.45')
        123.45
        >>> exercises.parse_float('invalid')
        0.0
        
        Args:
            number_string: String containing a number
            
        Returns:
            Parsed float value, or 0.0 if parsing fails
        """
        raise NotImplementedError("Method not implemented yet")

    def is_valid_number(self, number_string: str) -> bool:
        """
        Validate if a string represents a valid number.
        Check if the string can be parsed as a number.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.is_valid_number('123.45')
        True
        >>> exercises.is_valid_number('1,234')
        True
        >>> exercises.is_valid_number('abc')
        False
        >>> exercises.is_valid_number('12.34.56')
        False
        >>> exercises.is_valid_number('')
        False
        
        Args:
            number_string: String to validate
            
        Returns:
            True if string represents a valid number
        """
        raise NotImplementedError("Method not implemented yet")

    def format_file_size(self, bytes_size: int) -> str:
        """
        Format a file size in bytes using appropriate units (B, KB, MB, GB, TB).
        Use 1024 as the conversion factor and show 1 decimal place for values >= 1.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_file_size(512)
        '512 B'
        >>> exercises.format_file_size(1536)
        '1.5 KB'
        >>> exercises.format_file_size(1048576)
        '1.0 MB'
        >>> exercises.format_file_size(1073741824)
        '1.0 GB'
        >>> exercises.format_file_size(1099511627776)
        '1.0 TB'
        
        Args:
            bytes_size: File size in bytes
            
        Returns:
            Formatted file size string with appropriate unit
        """
        raise NotImplementedError("Method not implemented yet")

    def format_duration(self, milliseconds: int) -> str:
        """
        Format a duration in milliseconds as a readable time string.
        Show hours, minutes, seconds, and milliseconds as appropriate.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_duration(1500)
        '1.500s'
        >>> exercises.format_duration(65000)
        '1m 5s'
        >>> exercises.format_duration(3661000)
        '1h 1m 1s'
        >>> exercises.format_duration(90061500)
        '25h 1m 1.500s'
        
        Args:
            milliseconds: Duration in milliseconds
            
        Returns:
            Formatted duration string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_with_pattern(self, number: float, pattern: str) -> str:
        """
        Create a custom number format using Python's string formatting.
        Apply the pattern to format the number appropriately.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_with_pattern(1234.5, '{:,.2f}')
        '1,234.50'
        >>> exercises.format_with_pattern(0.75, '{:.1%}')
        '75.0%'
        >>> exercises.format_with_pattern(1234567, '{:.2e}')
        '1.23e+06'
        >>> exercises.format_with_pattern(42, '{:03d}')
        '042'
        
        Args:
            number: Number to format
            pattern: Python format string pattern
            
        Returns:
            Formatted number string using the pattern
        """
        raise NotImplementedError("Method not implemented yet")

    def format_as_ordinal(self, number: int) -> str:
        """
        Format a number as an ordinal (1st, 2nd, 3rd, 4th, etc.).
        Add the appropriate suffix based on the number.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_as_ordinal(1)
        '1st'
        >>> exercises.format_as_ordinal(2)
        '2nd'
        >>> exercises.format_as_ordinal(3)
        '3rd'
        >>> exercises.format_as_ordinal(4)
        '4th'
        >>> exercises.format_as_ordinal(21)
        '21st'
        >>> exercises.format_as_ordinal(22)
        '22nd'
        >>> exercises.format_as_ordinal(23)
        '23rd'
        >>> exercises.format_as_ordinal(11)
        '11th'
        >>> exercises.format_as_ordinal(12)
        '12th'
        >>> exercises.format_as_ordinal(13)
        '13th'
        
        Args:
            number: Number to format as ordinal
            
        Returns:
            Ordinal number string
        """
        raise NotImplementedError("Method not implemented yet")

    def format_in_base(self, number: int, base: int) -> str:
        """
        Format a number in different bases (binary, octal, hexadecimal).
        Convert the number to the specified base and format appropriately.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_in_base(255, 2)
        '11111111'
        >>> exercises.format_in_base(255, 8)
        '377'
        >>> exercises.format_in_base(255, 16)
        'FF'
        >>> exercises.format_in_base(42, 2)
        '101010'
        >>> exercises.format_in_base(42, 16)
        '2A'
        
        Args:
            number: Number to convert
            base: Target base (2, 8, 10, 16)
            
        Returns:
            Number string in the specified base
        """
        raise NotImplementedError("Method not implemented yet")

    def format_temperature(self, temperature: float, unit: str) -> str:
        """
        Format a temperature with the appropriate unit and precision.
        Support Celsius, Fahrenheit, and Kelvin units.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.format_temperature(25.5, 'C')
        '25.5°C'
        >>> exercises.format_temperature(77.9, 'F')
        '77.9°F'
        >>> exercises.format_temperature(298.65, 'K')
        '298.7K'
        >>> exercises.format_temperature(0, 'C')
        '0.0°C'
        
        Args:
            temperature: Temperature value
            unit: Unit ('C' for Celsius, 'F' for Fahrenheit, 'K' for Kelvin)
            
        Returns:
            Formatted temperature string
        """
        raise NotImplementedError("Method not implemented yet")

    def round_to_significant_digits(self, number: float, significant_digits: int) -> float:
        """
        Round a number to a specific number of significant digits.
        Use appropriate rounding to maintain the specified precision.
        
        Examples:
        >>> exercises = NumberFormattingExercises()
        >>> exercises.round_to_significant_digits(123.456, 4)
        123.5
        >>> exercises.round_to_significant_digits(0.00123456, 3)
        0.00123
        >>> exercises.round_to_significant_digits(1234567, 3)
        1230000.0
        >>> exercises.round_to_significant_digits(9.999, 2)
        10.0
        
        Args:
            number: Number to round
            significant_digits: Number of significant digits to keep
            
        Returns:
            Rounded number
        """
        raise NotImplementedError("Method not implemented yet")

File: java/src/main/java/com/zipcode/exercises/task09/NumberFormattingExercises.java
package com.zipcode.exercises.task09;

import java.util.Locale;

/**
 * Task 9: Number Formatting Exercises
 * 
 * This class contains exercises for learning number formatting operations in Java.
 * Students will implement methods that demonstrate:
 * - Decimal formatting with specific precision
 * - Currency formatting for different locales
 * - Percentage formatting and calculations
 * - Scientific notation formatting
 * - Number parsing and validation
 * - Custom number format patterns
 * 
 * Learning Objectives:
 * - Master DecimalFormat and NumberFormat classes
 * - Understand locale-specific number formatting
 * - Learn to format numbers for different contexts
 * - Practice number parsing and validation
 * - Handle different number formats and patterns
 * - Work with currency and percentage representations
 * 
 * @author ZipCode Cohort
 */
public class NumberFormattingExercises {

    /**
     * Format a decimal number with a specific number of decimal places.
     * Use DecimalFormat to control precision.
     * 
     * Examples:
     * formatDecimal(123.456789, 2) → "123.46"
     * formatDecimal(7.1, 3) → "7.100"
     * formatDecimal(42, 0) → "42"
     * formatDecimal(3.999, 1) → "4.0"
     * 
     * @param number Number to format
     * @param decimalPlaces Number of decimal places to show
     * @return Formatted decimal string
     */
    public String formatDecimal(double number, int decimalPlaces) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number with thousands separators (commas).
     * Use appropriate formatting to add commas for readability.
     * 
     * Examples:
     * formatWithThousandsSeparator(1234567.89) → "1,234,567.89"
     * formatWithThousandsSeparator(1000) → "1,000"
     * formatWithThousandsSeparator(42.5) → "42.5"
     * formatWithThousandsSeparator(999) → "999"
     * 
     * @param number Number to format
     * @return Number string with thousands separators
     */
    public String formatWithThousandsSeparator(double number) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number as currency using the default locale.
     * Use NumberFormat.getCurrencyInstance() for proper currency formatting.
     * 
     * Examples:
     * formatAsCurrency(123.45) → "$123.45" (US locale)
     * formatAsCurrency(0.99) → "$0.99"
     * formatAsCurrency(1000) → "$1,000.00"
     * formatAsCurrency(0) → "$0.00"
     * 
     * @param amount Amount to format as currency
     * @return Formatted currency string
     */
    public String formatAsCurrency(double amount) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number as currency using a specific locale.
     * Use NumberFormat.getCurrencyInstance(locale) for locale-specific formatting.
     * 
     * Examples:
     * formatAsCurrencyWithLocale(123.45, Locale.US) → "$123.45"
     * formatAsCurrencyWithLocale(123.45, Locale.FRANCE) → "123,45 €"
     * formatAsCurrencyWithLocale(123.45, Locale.JAPAN) → "￥123"
     * 
     * @param amount Amount to format as currency
     * @param locale Locale to use for formatting
     * @return Formatted currency string in the specified locale
     */
    public String formatAsCurrencyWithLocale(double amount, Locale locale) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a decimal value as a percentage.
     * Convert decimal to percentage (multiply by 100) and add % symbol.
     * 
     * Examples:
     * formatAsPercentage(0.1234, 2) → "12.34%"
     * formatAsPercentage(0.5, 1) → "50.0%"
     * formatAsPercentage(1.0, 0) → "100%"
     * formatAsPercentage(0.001, 3) → "0.100%"
     * 
     * @param value Decimal value (0.5 = 50%)
     * @param decimalPlaces Number of decimal places to show
     * @return Formatted percentage string
     */
    public String formatAsPercentage(double value, int decimalPlaces) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number in scientific notation.
     * Use exponential notation with specified decimal places.
     * 
     * Examples:
     * formatInScientificNotation(1234567, 2) → "1.23E6"
     * formatInScientificNotation(0.000123, 3) → "1.230E-4"
     * formatInScientificNotation(42, 1) → "4.2E1"
     * formatInScientificNotation(1, 0) → "1E0"
     * 
     * @param number Number to format
     * @param decimalPlaces Number of decimal places in mantissa
     * @return Scientific notation string
     */
    public String formatInScientificNotation(double number, int decimalPlaces) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number with leading zeros to a specific total width.
     * Use DecimalFormat or String formatting to add leading zeros.
     * 
     * Examples:
     * formatWithLeadingZeros(42, 5) → "00042"
     * formatWithLeadingZeros(123, 6) → "000123"
     * formatWithLeadingZeros(7, 3) → "007"
     * formatWithLeadingZeros(1000, 3) → "1000" (no truncation)
     * 
     * @param number Integer to format
     * @param totalWidth Desired total width
     * @return Number string with leading zeros
     */
    public String formatWithLeadingZeros(int number, int totalWidth) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Parse a string to extract a double value.
     * Handle various number formats and return the parsed value.
     * Return 0.0 if parsing fails.
     * 
     * Examples:
     * parseDouble("123.45") → 123.45
     * parseDouble("1,234.56") → 1234.56
     * parseDouble("$123.45") → 123.45
     * parseDouble("invalid") → 0.0
     * 
     * @param numberString String containing a number
     * @return Parsed double value, or 0.0 if parsing fails
     */
    public double parseDouble(String numberString) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Validate if a string represents a valid number.
     * Check if the string can be parsed as a number.
     * 
     * Examples:
     * isValidNumber("123.45") → true
     * isValidNumber("1,234") → true
     * isValidNumber("abc") → false
     * isValidNumber("12.34.56") → false
     * isValidNumber("") → false
     * 
     * @param numberString String to validate
     * @return True if string represents a valid number
     */
    public boolean isValidNumber(String numberString) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a file size in bytes using appropriate units (B, KB, MB, GB, TB).
     * Use 1024 as the conversion factor and show 1 decimal place for values >= 1.
     * 
     * Examples:
     * formatFileSize(512) → "512 B"
     * formatFileSize(1536) → "1.5 KB"
     * formatFileSize(1048576) → "1.0 MB"
     * formatFileSize(1073741824) → "1.0 GB"
     * formatFileSize(1099511627776L) → "1.0 TB"
     * 
     * @param bytes File size in bytes
     * @return Formatted file size string with appropriate unit
     */
    public String formatFileSize(long bytes) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a duration in milliseconds as a readable time string.
     * Show hours, minutes, seconds, and milliseconds as appropriate.
     * 
     * Examples:
     * formatDuration(1500) → "1.500s"
     * formatDuration(65000) → "1m 5s"
     * formatDuration(3661000) → "1h 1m 1s"
     * formatDuration(90061500) → "25h 1m 1.500s"
     * 
     * @param milliseconds Duration in milliseconds
     * @return Formatted duration string
     */
    public String formatDuration(long milliseconds) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Create a custom number format pattern and apply it to a number.
     * Use DecimalFormat with the provided pattern.
     * 
     * Examples:
     * formatWithPattern(1234.5, "#,##0.00") → "1,234.50"
     * formatWithPattern(0.75, "#0.0%") → "75.0%"
     * formatWithPattern(1234567, "#.##E0") → "1.23E6"
     * formatWithPattern(42, "000") → "042"
     * 
     * @param number Number to format
     * @param pattern DecimalFormat pattern string
     * @return Formatted number string using the pattern
     */
    public String formatWithPattern(double number, String pattern) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number as an ordinal (1st, 2nd, 3rd, 4th, etc.).
     * Add the appropriate suffix based on the number.
     * 
     * Examples:
     * formatAsOrdinal(1) → "1st"
     * formatAsOrdinal(2) → "2nd"
     * formatAsOrdinal(3) → "3rd"
     * formatAsOrdinal(4) → "4th"
     * formatAsOrdinal(21) → "21st"
     * formatAsOrdinal(22) → "22nd"
     * formatAsOrdinal(23) → "23rd"
     * formatAsOrdinal(11) → "11th"
     * formatAsOrdinal(12) → "12th"
     * formatAsOrdinal(13) → "13th"
     * 
     * @param number Number to format as ordinal
     * @return Ordinal number string
     */
    public String formatAsOrdinal(int number) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a number in different bases (binary, octal, hexadecimal).
     * Convert the number to the specified base and format appropriately.
     * 
     * Examples:
     * formatInBase(255, 2) → "11111111"
     * formatInBase(255, 8) → "377"
     * formatInBase(255, 16) → "FF"
     * formatInBase(42, 2) → "101010"
     * formatInBase(42, 16) → "2A"
     * 
     * @param number Number to convert
     * @param base Target base (2, 8, 10, 16)
     * @return Number string in the specified base
     */
    public String formatInBase(int number, int base) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Format a temperature with the appropriate unit and precision.
     * Support Celsius, Fahrenheit, and Kelvin units.
     * 
     * Examples:
     * formatTemperature(25.5, "C") → "25.5°C"
     * formatTemperature(77.9, "F") → "77.9°F"
     * formatTemperature(298.65, "K") → "298.7K"
     * formatTemperature(0, "C") → "0.0°C"
     * 
     * @param temperature Temperature value
     * @param unit Unit ("C" for Celsius, "F" for Fahrenheit, "K" for Kelvin)
     * @return Formatted temperature string
     */
    public String formatTemperature(double temperature, String unit) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Round a number to a specific number of significant digits.
     * Use appropriate rounding to maintain the specified precision.
     * 
     * Examples:
     * roundToSignificantDigits(123.456, 4) → 123.5
     * roundToSignificantDigits(0.00123456, 3) → 0.00123
     * roundToSignificantDigits(1234567, 3) → 1230000.0
     * roundToSignificantDigits(9.999, 2) → 10.0
     * 
     * @param number Number to round
     * @param significantDigits Number of significant digits to keep
     * @return Rounded number
     */
    public double roundToSignificantDigits(double number, int significantDigits) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }
}
